File: app/src/main/java/us/hnry/fancy/fragments/MainFragment.java
package us.hnry.fancy.fragments;

import android.app.Fragment;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.concurrent.ExecutionException;

import us.hnry.fancy.R;
import us.hnry.fancy.SearchActivity;
import us.hnry.fancy.adapters.StockRecycler;
import us.hnry.fancy.data.FetchStockTask;
import us.hnry.fancy.models.Stock;
import us.hnry.fancy.utils.QuoteQueryBuilder;
import us.hnry.fancy.utils.Utility;

/**
 * Created by Henry on 2/6/2016.
 */
public class MainFragment extends Fragment {

    private RecyclerView mRecyclerView;
    private FloatingActionButton fab;
    private ArrayList<Stock> mQuotes;
    private SharedPreferences preferences;
    private SharedPreferences.Editor editor;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View layout = inflater.inflate(R.layout.fragment_main_recycler, container, false);

        fab = (FloatingActionButton) getActivity().findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(getActivity(), SearchActivity.class).putExtra(Utility.SEARCH_INTENT, Utility.THOR_SEARCH));
            }
        });

        mRecyclerView = (RecyclerView) layout.findViewById(R.id.fragment_main_recycler_view);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        return layout;
    }

    @Override
    public void onResume() {
        super.onResume();
        refreshMain();
    }

    public void refreshMain() {
        if (preferences == null) {
            preferences = getActivity().getSharedPreferences(Utility.PERSISTENT, Context.MODE_PRIVATE);
            editor = preferences.edit();
        }

        final ProgressDialog progressDialog = new ProgressDialog(getActivity());
        progressDialog.setMessage("Refreshing your data");
        progressDialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
        progressDialog.setCancelable(false);
        progressDialog.setIndeterminate(true);
        progressDialog.show();

        //Instantiate the async task
        final FetchStockTask task = new FetchStockTask(getActivity());

        final String[] symbolsToQuery =
                Utility.getSymbols(
                        preferences.getStringSet(Utility.PERSISTENT_SYMBOLS_SET,
                                new HashSet<>(Arrays.asList(Utility.DEFAULT_SYMBOLS))));

        final QuoteQueryBuilder[] queryBuilder = {null};
        new Thread() {
            @Override
            public void run() {
                queryBuilder[0] = new QuoteQueryBuilder(symbolsToQuery);
                //Execute the task
                task.execute(queryBuilder[0].build());
                try {
                    //Fetch the result of the background thread
                    mQuotes = task.get();
                    if (mQuotes != null) {
                        final StockRecycler adapter = new StockRecycler(mQuotes, getActivity());

                        getActivity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                //Set the adapter to the list view
                                mRecyclerView.setAdapter(adapter);
                                editor.clear();
                                editor.putStringSet(Utility.PERSISTENT_SYMBOLS_SET,
                                        new HashSet<>(Arrays.asList(symbolsToQuery)));
                                editor.apply();
                                progressDialog.dismiss();
                            }
                        });
                    }
                } catch (InterruptedException | ExecutionException e) {
                    e.printStackTrace();
                }
            }
        }.start();
    }
}

File: app/src/main/java/us/hnry/fancy/fragments/MainFragment.kt
package us.hnry.fancy.fragments

import android.app.Fragment
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.firebase.crash.FirebaseCrash
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import us.hnry.fancy.R
import us.hnry.fancy.adapters.QuotesAdapter
import us.hnry.fancy.data.StockRepositoryImpl
import us.hnry.fancy.network.StockServiceImpl
import us.hnry.fancy.network.model.Symbol
import us.hnry.fancy.presentation.StockPresenter
import us.hnry.fancy.presentation.StockPresenterImpl
import us.hnry.fancy.presentation.model.StockDetail
import us.hnry.fancy.presentation.view.StockView
import us.hnry.fancy.ui.ItemTouchCallback
import us.hnry.fancy.utils.SymbolsHelper
import java.util.concurrent.Executors

/**
 * @author Henry
 * 10/8/2017
 */
class MainFragment : Fragment(), StockView, PersistentSymbolsChangedListener {
    private val symbolsHelper by lazy { SymbolsHelper(activity, this) }

    private lateinit var presenter: StockPresenter
    private lateinit var recyclerView: RecyclerView

    private lateinit var adapter: QuotesAdapter

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View {
        retainInstance = true

        val layout: View = inflater!!.inflate(R.layout.fragment_main_recycler, container, false)
        recyclerView = layout.findViewById(R.id.fragment_main_recycler_view)

        initPresenter()
        adapter = QuotesAdapter(activity, this)
        recyclerView.adapter = adapter

        recyclerView.layoutManager = LinearLayoutManager(activity)
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.setHasFixedSize(true)

        val callback: ItemTouchHelper.Callback = ItemTouchCallback(adapter)
        val itemTouchHelper = ItemTouchHelper(callback)
        itemTouchHelper.attachToRecyclerView(recyclerView)
        return layout
    }

    private fun initPresenter() {
        presenter = StockPresenterImpl(
                StockRepositoryImpl(StockServiceImpl.get()), AndroidSchedulers.mainThread(),
                Schedulers.from(Executors.newSingleThreadExecutor()), *symbolsHelper.getPersistentSymbolsSetAsArray()
        )
    }

    override fun onResume() {
        attachPresenter()
        super.onResume()
    }

    override fun onStop() {
        detachPresenter()
        super.onStop()
    }

    override fun attachPresenter() {
        presenter.attachView(this)
    }

    override fun detachPresenter() {
        presenter.detachView()
    }

    override fun displayStockData(listOfQuotes: List<StockDetail>) {
        adapter.swapList(listOfQuotes)
    }

    override fun logMessage(message: String) {
        FirebaseCrash.log(message)
    }

    override fun onSymbolAdded(symbol: Symbol) {
        presenter.symbolSetChanged(*symbolsHelper.getPersistentSymbolsSetAsArray())

        recyclerView.let {
            Snackbar.make(it, "${symbol.symbol} will be visible after the next update",
                    Snackbar.LENGTH_SHORT).show()
        }
    }

    override fun onSymbolRemoved(symbol: Symbol) {
        presenter.symbolSetChanged(*symbolsHelper.getPersistentSymbolsSetAsArray())

        recyclerView.let {
            Snackbar.make(it, "Symbol removed from Favorites", Snackbar.LENGTH_LONG)
                    .setAction("Undo") { symbolsHelper.addSymbol(symbol) }
                    .show()
        }
    }
}